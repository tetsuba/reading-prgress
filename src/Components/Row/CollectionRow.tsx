import * as R from 'ramda'
import { useDispatch } from 'react-redux'

// TYPES
import { CollectionsTypes } from '../../store/selector.types'

// STORE
import { updateCurrentCollectionId } from '../../store/current/currentSlice'

// COMPONENTS
import Row from './Row'
import Button from '../Button/Button'

export type CollectionPropTypes = {
    data?: CollectionsTypes | undefined
    index?: number
}

export default function CollectionRow(props: CollectionPropTypes) {
    const dispatch = useDispatch()
    const { data, index = 0 } = props

    if (R.isNil(data)) return <>loading...</>

    const text = `${data.title} (${data.numberOfBooks})`
    const iconColour = data.completed ? 'text-green-500' : ''

    return (
        <Row index={index} text={text} iconColor={iconColour} icon="library">
            <Button
                className="flex items-center"
                data-testid="collection-button"
                icon="eye"
                template={`${data.completed ? 'disabled' : 'secondary'}`}
                onClick={() => dispatch(updateCurrentCollectionId(data.id))}
                right
            >
                <span className="mr-1 hidden md:inline">View Books</span>
            </Button>
        </Row>
    )
}
