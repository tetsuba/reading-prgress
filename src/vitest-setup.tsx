import { QueryClient, QueryClientProvider } from 'react-query'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from './store/store'
import { updateUser } from './store/user/userSlice'
import { addBooks } from './store/books/booksSlice'
import { addStudents } from './store/students/studentsSlice'
import { studentsMockData } from '../tests/mockData/students'
import booksMockData from '../tests/mockData/books'
import { userMockData } from '../tests/mockData/user'

const portal = document.createElement('div')
portal.setAttribute('id', 'modal')
document.body.appendChild(portal)

const mockResults = {
    results: [[{ transcript: 'This is a story' }]]
}

class MockSpeech {
    start() {
        // @ts-ignore
        this.onresult(mockResults)
    }
    stop = vi.fn()
    onresult = vi.fn()
    onend = vi.fn()
}

Object.defineProperty(window, 'webkitSpeechRecognition', {
    writable: true,
    value: vi.fn().mockImplementation(() => new MockSpeech())
})

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            staleTime: 0,
            cacheTime: 0,
            retryDelay: 1,
            retry: 0
        }
    }
})

type PropTypes = {
    pathname: string
    children: JSX.Element
    notAuthenticated?: boolean
}

export function updateStoreWithMockData() {
    store.dispatch(updateUser(userMockData))
    store.dispatch(addBooks(booksMockData))
    store.dispatch(addStudents(studentsMockData))
}

export function WrapperWith_Store_Query_Router(props: PropTypes) {
    if (!props.notAuthenticated) updateStoreWithMockData()
    return (
        <Provider store={store}>
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={[props.pathname]}>
                    {props.children}
                </MemoryRouter>
            </QueryClientProvider>
        </Provider>
    )
}

export function WrapperWith_Store_Router(props: PropTypes) {
    return (
        <Provider store={store}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </Provider>
    )
}

export function WrapperWithQueryAndRouter(props: PropTypes) {
    return (
        <QueryClientProvider client={queryClient}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </QueryClientProvider>
    )
}

export function WrapperWithQuery(props: { children: JSX.Element }) {
    return (
        <QueryClientProvider client={queryClient}>
            {props.children}
        </QueryClientProvider>
    )
}

export function WrapperWithRouter(props: PropTypes) {
    return (
        <MemoryRouter initialEntries={[props.pathname]}>
            {props.children}
        </MemoryRouter>
    )
}
