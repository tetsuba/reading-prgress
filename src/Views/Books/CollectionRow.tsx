import { ApiCollectionTypes } from '../../lib/service-types'
import { useDispatch } from 'react-redux'
import Row from '../../Components/Row/Row'
import { ViewBooksButton } from '../../Components/Button/Buttons'
import { updateViewBookCollection } from '../../store/view/viewSlice'

type CollectionPropTypes = {
    data?: ApiCollectionTypes | undefined
    index?: number
}

export default function CollectionRow(props: CollectionPropTypes) {
    const dispatch = useDispatch()
    const { data, index = 0 } = props
    const text = props.data
        ? `${props.data.title} (${props.data.books.length})`
        : ''
    return (
        <Row index={index} text={text} iconColor="" icon="library">
            <ViewBooksButton
                onClick={() => data && dispatch(updateViewBookCollection(data))}
            />
        </Row>
    )
}
