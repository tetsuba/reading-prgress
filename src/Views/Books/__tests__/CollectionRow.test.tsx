import {render} from "@testing-library/react";
import {WrapperWith_Store_Query_Router} from "../../../vitest-setup";
import CollectionRow from "../CollectionRow";


describe('CollectionRow', () => {
    test('props data is undefined', () => {
        const { asFragment } = render(
            <WrapperWith_Store_Query_Router pathname={'/books'}>
                <CollectionRow data={undefined} index={0} />
            </WrapperWith_Store_Query_Router>
        )
        expect(asFragment()).toMatchSnapshot()
    })
})
