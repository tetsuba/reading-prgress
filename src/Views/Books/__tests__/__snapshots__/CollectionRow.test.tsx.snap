// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CollectionRow > if collection row icon is green 1`] = `
<DocumentFragment>
  <div
    class="flex justify-between px-4 py-5 sm:gap-4 sm:px-6 bg-gray-50"
  >
    <div
      class="flex items-center"
    >
      <span
        class="mr-6 text-green-500"
      >
        <svg
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 21v-8.25M15.75 21v-8.25M8.25 21v-8.25M3 9l9-6 9 6m-1.5 12V10.332A48.36 48.36 0 0012 9.75c-2.551 0-5.056.2-7.5.582V21M3 21h18M12 6.75h.008v.008H12V6.75z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
      <span
        class="font-medium text-gray-800"
      >
        title collection 1 (2)
      </span>
    </div>
    <div
      class="flex justify-end"
    >
      <button
        class="flex items-center rounded-lg font-medium text-base text-center px-5 py-2.5 text-gray-400 bg-gray-200 flex items-center"
        data-testid="collection-button"
      >
        <span
          class="mr-1 hidden md:inline"
        >
          View Books
        </span>
        <svg
          class="h-5 w-5 "
          fill="none"
          stroke="currentColor"
          stroke-width="1"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2.036 12.322a1.012 1.012 0 010-.639C3.423 7.51 7.36 4.5 12 4.5c4.638 0 8.573 3.007 9.963 7.178.07.207.07.431 0 .639C20.577 16.49 16.64 19.5 12 19.5c-4.638 0-8.573-3.007-9.963-7.178z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
          <path
            d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
         
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CollectionRow > props data is undefined 1`] = `
<DocumentFragment>
  loading...
</DocumentFragment>
`;
